staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/matcher.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generic
+
+import (
+	"context"
+
+	"github.com/google/cel-go/cel"
+	celtypes "github.com/google/cel-go/common/types"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/admission"
+	admissionmetrics "k8s.io/apiserver/pkg/admission/metrics"
+	celplugin "k8s.io/apiserver/pkg/admission/plugin/cel"
+	celconfig "k8s.io/apiserver/pkg/apis/cel"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/klog/v2"
+)
+
+var _ celplugin.ExpressionAccessor = &MatchCondition{}
+
+// MatchCondition contains the inputs needed to compile, evaluate and match a cel expression
+type MatchCondition struct {
+	Expression      string
+	Name            string
+	WebhookStepType string
+}
+
+func (v *MatchCondition) GetExpression() string {
+	return v.Expression
+}
+
+func (v *MatchCondition) ReturnTypes() []*cel.Type {
+	return []*cel.Type{cel.BoolType}
+}
+
+var _ Matcher = &matcher{}
+
+// matcher evaluates compiled cel expressions and determines if they match the given request or not
+type matcher struct {
+	filter     celplugin.Filter
+	authorizer authorizer.Authorizer
+}
+
+func NewMatcher(filter celplugin.Filter, authorizer authorizer.Authorizer) Matcher {
+	return &matcher{
+		filter:     filter,
+		authorizer: authorizer,
+	}
+}
+
+// Match takes a list of Evaluation converts them into actionable a decision of whether they match or not.
+// If no match returns name of failed condition, versionedParams can be nil for no additional params
+func (m *matcher) Match(ctx context.Context, versionedAttr *admission.VersionedAttributes, versionedParams runtime.Object) (bool, string, error) {
+	evalResults, err := m.filter.ForInput(ctx, versionedAttr, celplugin.CreateAdmissionRequest(versionedAttr.Attributes), celplugin.OptionalVariableBindings{
+		VersionedParams: versionedParams,
+		Authorizer:      m.authorizer,
+	}, celconfig.PerCallLimit)
```

Comment:
Here we might need to have a separate runtimeCostBudget for matchConditions. 

We were using `celconfig.RuntimeCELCostBudget` for validations in ValidatingAdmissionPolicy which roughly will allow 10 expensive expressions or 20 authz checks. 

Just offline discussed with @jpbetz . For matchConditions, we are thinking the initial number to be 1/4(2500000) which will allow 5 authz check or 2.5 super expensive expressions (we could always start low and increase later if needed for backward compatibility. Curious what others think?
cc @tallclair @maxsmythe

Reply:
This is per-expression, right? I think the more important limit is across all expressions. 5 authz checks seems like more than enough per-expression, and is probably sufficient for total as well. I agree with low-balling it to start.

Reply:
My intent was for the 2500000 to be the limit across all expressions (it would be 1/4th celconfig.RuntimeCELCostBudget which is what we use for the limit elsewhere).  My rationale for the 1/4 is that these are just preconditions used to decide if something else should be done, so they shouldn't be super expensive.

Note that there is also a per-expression limit in play here, it's at 1000000, which is enough for 2 authz checks.

Reply:
I created a new constant `RuntimeCELCostBudgetMatchConditions` with value 2500000